sonixflasher.c

```diff
@@ -401,142 +807,84 @@ int main(int argc, char* argv[])
         handle = hid_open(vid, pid, NULL);
         attempt_no++;
     }
 
-    if(handle)
-    {
+    if (handle) {
+        printf("\n");
         printf("Device opened successfully...\n");
 
         // Check VID/PID
-        if(vid != SONIX_VID  || (pid != SN248_PID && pid != SN248B_PID && pid != SN268_PID))
-        {
-            if(vid == EVISION_VID && !reboot_requested) printf("Warning: eVision VID detected! You probably need to use the reboot option.\n");
-            if(vid == APPLE_VID && !reboot_requested) printf("Warning: Apple VID detected! You probably need to use the reboot option.\n");
+        if (vid != SONIX_VID || (pid != SN248_PID && pid != SN248B_PID && pid != SN248C_PID && pid != SN268_PID && pid != SN289_PID && pid != SN299_PID)) {
+            if (vid == EVISION_VID && !reboot_requested) printf("Warning: eVision VID detected! You probably need to use the reboot option.\n");
+            if (vid == APPLE_VID && !reboot_requested) printf("Warning: Apple VID detected! You probably need to use the reboot option.\n");
             printf("Warning: Flashing a non-sonix bootloader device, you are now on your own.\n");
             sleep(3);
-
-            // Set max firmware to 64k, useful when flashing a Sonix Board that isnt in BL mode (Redragons, Keychrons)
-            MAX_FIRMWARE = MAX_FIRMWARE_SN32F240; // Maybe add a param to override this (?)
-            printf("Warning: We assume a ROM size of 64k.\n");
+        }
+        attempt_no = 1;
+        bool ok    = protocol_init(handle, reboot_requested, reboot_opt);
+        while (!ok && attempt_no <= MAX_ATTEMPTS) {
+            printf("Device failed to init, re-trying in 3 seconds. Attempt %d of %d...\n", attempt_no, MAX_ATTEMPTS);
             sleep(3);
+            ok = protocol_init(handle, reboot_requested, reboot_opt);
+            attempt_no++;
         }
-
-        // Set max fw size depending on VID/PID
-        if(vid == SONIX_VID)
-        {
-            switch (pid)
-            {
-                case SN248_PID:
-                case SN248B_PID:
-                    MAX_FIRMWARE = MAX_FIRMWARE_SN32F240;
-                    break;
-
-                case SN268_PID:
-                    MAX_FIRMWARE = MAX_FIRMWARE_SN32F260;
-
-                    if(!flash_jumploader && offset == 0) // Failsafe when flashing a 268 w/o jumploader and offset
-                    {
-                        printf("Warning: Flashing 26X without offset.\n");
-                        printf("Fail safing to offset 0x%04x\n", QMK_OFFSET_DEFAULT);
-
-                        offset = QMK_OFFSET_DEFAULT;
-                    }
-                    break;
-
-                default:
-                    fprintf(stderr, "ERROR: Unsupported sonix bootloader device. Quitting.\n");
-                    exit(1);
-            }
+        if (!ok) error(handle);
+        sleep(3);
+        if (chip == SN240 || chip == SN290) ok = protocol_code_option_check(handle); // 240 and 290
```

Comment:
Are you sure about these version checks? The checks I saw where for != 268 and != 248b which leaves quite a few more models than just 240 and 290.

Reply:
no really sure about those, this should be updated indeed